packages/language-service/src/lib-new/lang-service-context.ts

```diff
@@ -0,0 +1,203 @@
+    }
+    public getPosition(offset: number = this.offset) {
+        return this.document.positionAt(offset);
+    }
+    public isInRoot() {
+        return this.location.base.node.type === 'root';
+    }
+    public isInSelector() {
+        return !!this.location.selector;
+    }
+    public getSelectorContext() {
+        const nestedSelectors = this.collectSelector();
+        // collect the CSS resolve chain for each selector node
+        const selectorAstResolveMap = new Map<ImmutableSelectorNode, CSSResolve[]>();
+        let currentAnchor = undefined;
+        for (const selector of nestedSelectors) {
+            // ToDo(tech-debt): provide an internal selector resolve without the transformer
+            const transformer = (this.stylable as any).createTransformer() as StylableTransformer;
+            const selectorContext = transformer.createSelectorContext(
+                this.meta,
+                selector as SelectorList, // doesn't mutate due to `selectorContext.transform = false`
+                postcss.rule({ selector: stringifySelectorAst(selector) })
+            );
+            if (currentAnchor) {
+                selectorContext.currentAnchor = currentAnchor;
+                selectorContext.nestingSelectorAnchor = currentAnchor;
+            }
+            selectorContext.transform = false;
+            selectorContext.selectorAstResolveMap = selectorAstResolveMap;
+            transformer.scopeSelectorAst(selectorContext);
+            // ToDo: handle multiple selectors intersection
+            currentAnchor = selectorContext.currentAnchor;
+        }
+        // reference interest points
+        const locationSelector = this.location.selector!;
+        const nodeAtCursor = locationSelector.node;
+        const resolvedSelectorChain = this.aggregateResolvedChain(
+            nestedSelectors,
+            selectorAstResolveMap
+        );
+        return {
+            nodeAtCursor: locationSelector.node,
+            selectorAtCursor: stringifySelectorAst(nodeAtCursor).slice(
+                0,
+                this.location.selector!.offsetInNode
+            ),
+            fullSelectorAtCursor: stringifySelectorAst(
+                nestedSelectors[nestedSelectors.length - 1]
+            ).slice(0, nodeAtCursor.start + locationSelector.offsetInNode),
+            selectorAstResolveMap,
+            nestedSelectors,
+            resolvedSelectorChain,
+        };
+    }
+    /**
+     * Find selector not which caret is at and traverse back to collect
+     * resolved elements with any selector information after them.
+     */
+    private aggregateResolvedChain(
+        nestedSelectors: ImmutableSelectorList[],
+        selectorAstResolveMap: Map<ImmutableSelectorNode, CSSResolve[]>
+    ) {
+        const resolvedChain: ResolveChainItem[] = [];
+
+        let item: ResolveChainItem = {};
+        const nestingSelectors = [...nestedSelectors];
+        const caretSelector = nestingSelectors.pop()!;
```

Comment:
maybe `const [...nestingSelectors, caretSelector] = nestedSelectors`

Reply:
Don't you wish that would work right... after rust I forgot JavaScript don't support that:

`A rest element must be last in a destructuring pattern.ts(2462)` 😞